src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Contracts/ExecutionManager/Helpers/HotColdLookup.cs

```diff
@@ -0,0 +1,134 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace Microsoft.Diagnostics.DataContractReader.ExecutionManagerHelpers;
+
+internal sealed class HotColdLookup
+{
+    public static HotColdLookup Create(Target target)
+        => new HotColdLookup(target);
+
+    private readonly Target _target;
+
+    private HotColdLookup(Target target)
+    {
+        _target = target;
+    }
+
+    public uint GetHotFunctionIndex(uint numHotColdMap, TargetPointer hotColdMap, uint runtimeFunctionIndex)
+    {
+        if (!IsColdCode(numHotColdMap, hotColdMap, runtimeFunctionIndex))
+            return runtimeFunctionIndex;
+
+        uint hotIndex;
+        if (!TryLookupHotColdMappingForMethod(numHotColdMap, hotColdMap, runtimeFunctionIndex, out hotIndex, out _))
+            return runtimeFunctionIndex;
+
+        // If runtime function is in the cold part, get the associated hot part
+        Debug.Assert((hotIndex & 1) == 1);
+        return _target.Read<uint>(hotColdMap + (ulong)hotIndex * sizeof(uint));
```

Comment:
Do we want to drop the low order bit prior to the multiplication?

Reply:
Okay, so we are validating the number is odd, not that the lowest order bit it set. Adding a comment as to what that is asserting would clear this up.